### PR TITLE
Benchmark precompile op in gpu driver

### PR DESCRIPTION
**Pull request by pfultz2**

### CODE REVIEW

File: src/targets/gpu/driver/precompile_op.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2022 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (umangyadav) [inline src/targets/gpu/driver/precompile_op.cpp:4]:
```suggestion
 * Copyright (c) 2015-2023 Advanced Micro Devices, Inc. All rights reserved.
```
Until we batch-process all files to update the year to 2023, we can at least use 2023 for the new files.

**Review by umangyadav — APPROVED** [review]
LGTM

**Comment by causten** [comment]
Tidy problem, Performance drop

**Review by kahmed10 — APPROVED** [review]
LGTM. Please check tidy warnings.

**Comment by causten** [comment]
@pfultz2  I can merge once tidy problem is fixed

**Comment by causten** [comment]
```
[ 54%] Building CXX object test/CMakeFiles/header_src_include_migraphx_instruction_ref_hpp.dir/header-main-include-header_src_include_migraphx_instruction_ref_hpp.cpp.o
cd /home/jenkins/workspace/AMDMIGraphX_PR-2340/build/test && ccache /opt/rocm/llvm/bin/clang++ -DBUILD_DEV -DHAS_HALF_V1 -DHAVE_GPU -DMIGRAPHX_HAS_FIND_2_API -DMIGRAPHX_HAS_FIND_MODE_API -DMIGRAPHX_PREALLOCATE_MIOPEN_BUFFERS -DMIGRAPHX_USE_ROCBLAS_TUNING_API -DROCBLAS_BETA_FEATURES_API -DROCBLAS_NO_DEPRECATED_WARNINGS -D__HIP_PLATFORM_AMD__=1 -D__HIP_PLATFORM_HCC__=1 -I/home/jenkins/workspace/AMDMIGraphX_PR-2340/test/include -I/home/jenkins/workspace/AMDMIGraphX_PR-2340/src/include -I/home/jenkins/workspace/AMDMIGraphX_PR-2340/build/src/include -I/home/jenkins/workspace/AMDMIGraphX_PR-2340/build/src/onnx/include -I/home/jenkins/workspace/AMDMIGraphX_PR-2340/build/src/tf/include -I/home/jenkins/workspace/AMDMIGraphX_PR-2340/build/src/targets/ref/include -I/home/jenkins/workspace/AMDMIGraphX_PR-2340/src/targets/ref/include -I/home/jenkins/workspace/AMDMIGraphX_PR-2340/build/src/targets/gpu/include -I/home/jenkins/workspace/AMDMIGraphX_PR-2340/src/targets/gpu/include -isystem /usr/local/include/half -isystem /opt/rocm-5.7.0/include -isystem /opt/rocm/include -Werror -O3 -DNDEBUG -std=c++17 -Wall -Wextra -Wcomment -Wendif-labels -Wformat -Winit-self -Wreturn-type -Wsequence-point -Wswitch -Wtrigraphs -Wundef -Wuninitialized -Wunreachable-code -Wunused -Wno-sign-compare -Weverything -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-conversion -Wno-double-promotion -Wno-exit-time-destructors -Wno-extra-semi -Wno-extra-semi-stmt -Wno-float-conversion -Wno-gnu-anonymous-struct -Wno-gnu-zero-variadic-macro-arguments -Wno-missing-prototypes -Wno-nested-anon-types -Wno-option-ignored -Wno-padded -Wno-shorten-64-to-32 -Wno-sign-conversion -Wno-unused-command-line-argument -Wno-weak-vtables -Wno-c99-extensions -MD -MT test/CMakeFiles/header_src_include_migraphx_instruction_ref_hpp.dir/header-main-include-header_src_include_migraphx_instruction_ref_hpp.cpp.o -MF CMakeFiles/header_src_include_migraphx_instruction_ref_hpp.dir/header-main-include-header_src_include_migraphx_instruction_ref_hpp.cpp.o.d -o CMakeFiles/header_src_include_migraphx_instruction_ref_hpp.dir/header-main-include-header_src_include_migraphx_instruction_ref_hpp.cpp.o -c /home/jenkins/workspace/AMDMIGraphX_PR-2340/build/test/header-main-include-header_src_include_migraphx_instruction_ref_hpp.cpp
make  -f test/CMakeFiles/header_src_include_migraphx_iota_iterator_hpp.dir/build.make test/CMakeFiles/header_src_include_migraphx_iota_iterator_hpp.dir/depend
make[2]: Entering directory '/home/jenkins/workspace/AMDMIGraphX_PR-2340/build'
cd /home/jenkins/workspace/AMDMIGraphX_PR-2340/build && /usr/local/lib/python3.8/dist-packages/cmake/data/bin/cmake -E cmake_depends "Unix Makefiles" /home/jenkins/workspace/AMDMIGraphX_PR-2340 /home/jenkins/workspace/AMDMIGraphX_PR-2340/test /home/jenkins/workspace/AMDMIGraphX_PR-2340/build /home/jenkins/workspace/AMDMIGraphX_PR-2340/build/test /home/jenkins/workspace/AMDMIGraphX_PR-2340/build/test/CMakeFiles/header_src_include_migraphx_iota_iterator_hpp.dir/DependInfo.cmake --color=
/home/jenkins/workspace/AMDMIGraphX_PR-2340/src/targets/gpu/driver/precompile_op.cpp:50:13: error: no matching member function for call to 'add_instruction'
        mm->add_instruction(preop, args);
        ~~~~^~~~~~~~~~~~~~~
/home/jenkins/workspace/AMDMIGraphX_PR-2340/src/include/migraphx/module.hpp:81:21: note: candidate function not viable: 'this' argument has type 'const module', but method is not marked const
    instruction_ref add_instruction(const operation& op, std::vector<instruction_ref> args);
                    ^
/home/jenkins/workspace/AMDMIGraphX_PR-2340/src/include/migraphx/module.hpp:76:21: note: candidate template ignored: requirement '(migraphx::and_<false>{})' was not satisfied [with Ts = <std::vector<instruction_ref>>, PrivateRequires75 = 75]
    instruction_ref add_instruction(operation op, Ts... args)
                    ^
/home/jenkins/workspace/AMDMIGraphX_PR-2340/src/include/migraphx/module.hpp:83:21: note: candidate function not viable: requires 3 arguments, but 2 were provided
    instruction_ref add_instruction(const operation& op,
                    ^
/home/jenkins/workspace/AMDMIGraphX_PR-2340/src/targets/gpu/driver/precompile_op.cpp:48:20: error: 'this' argument to member function 'add_parameter' has type 'const module', but function is not marked const
            return mm->add_parameter("x" + std::to_string(i), input);
                   ^~
/usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_algo.h:4382:14: note: in instantiation of function template specialization 'migraphx::gpu::driver::precompile_op::create_preop_program(const operation &, std::vector<shape>)::(anonymous class)::operator()<migraphx::shape, long>' requested here
        *__result = __binary_op(*__first1, *__first2);
                    ^
/home/jenkins/workspace/AMDMIGraphX_PR-2340/src/include/migraphx/ranges.hpp:204:10: note: in instantiation of function template specialization 'std::transform<__gnu_cxx::__normal_iterator<migraphx::shape *, std::vector<migraphx::shape>>, migraphx::basic_iota_iterator<migraphx::id>, std::back_insert_iterator<std::vector<std::_List_iterator<migraphx::instruction>>>, (lambda at /home/jenkins/workspace/AMDMIGraphX_PR-2340/src/targets/gpu/driver/precompile_op.cpp:47:75)>' requested here
    std::transform(r1.begin(), r1.end(), r2.begin(), it, f);
         ^
/home/jenkins/workspace/AMDMIGraphX_PR-2340/src/targets/gpu/driver/precompile_op.cpp:47:9: note: in instantiation of function template specialization 'migraphx::transform<std::vector<migraphx::shape> &, migraphx::iterator_range<migraphx::basic_iota_iterator<migraphx::id>>, std::back_insert_iterator<std::vector<std::_List_iterator<migraphx::instruction>>>, (lambda at /home/jenkins/workspace/AMDMIGraphX_PR-2340/src/targets/gpu/driver/precompile_op.cpp:47:75)>' requested here
        transform(inputs, range(inputs.size()), std::back_inserter(args), [&](auto input, auto i) {
        ^
/home/jenkins/workspace/AMDMIGraphX_PR-2340/src/include/migraphx/module.hpp:161:21: note: 'add_parameter' declared here
    instruction_ref add_parameter(std::string name, shape s);
                    ^
```

**Pull request merged**